src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -97,11 +104,47 @@ private abstract class AbstractTableDataSourceFindUsagesContext :
                 Debug.Assert(_findReferencesWindow.Manager.Sources.Count == 0);
 
                 // And add ourselves as the source of results for the window.
-                _findReferencesWindow.Manager.AddSource(this);
+                // Additionally, add custom columns to display custom reference information.
+                _findReferencesWindow.Manager.AddSource(this, customColumns.SelectAsArray(c => c.Name));
 
                 // After adding us as the source, the manager should immediately call into us to
                 // tell us what the data sink is.
                 Debug.Assert(_tableDataSink != null);
+
+                // Initialize and update custom column states.
+                // Note that the TableControl.SetColumnStates is not dependent on order of the input column states.
+                _customColumnTitleToStatesMap = GetInitialCustomColumnStates(findReferencesWindow.TableControl.ColumnStates, customColumns);
+                TableControl.SetColumnStates(_customColumnTitleToStatesMap.Values);
+            }
+
+            private static Dictionary<string, ColumnState2> GetInitialCustomColumnStates(
+                IReadOnlyList<ColumnState> allColumnStates,
+                ImmutableArray<AbstractFindUsagesCustomColumnDefinition> customColumns)
+            {
+                var customColumnStatesMap = new Dictionary<string, ColumnState2>(customColumns.Length);
+                var customColumnNames = new HashSet<string>(customColumns.Select(c => c.Name));
+                
+                // Set the default visibility for each custom column.
+                // If there is an existing column state for the custom column, flip it to be non-visible by default at the start of FAR query.
+                // We do so because the column will have empty values for all results for a FAR query for certain cases such as types, literals, no references found case, etc.
+                // It is preferable to dynamically hide an empty column for such queries, and dynamically show the column if it has at least one non-default value.
+                foreach (ColumnState2 columnState in allColumnStates.Where(c => customColumnNames.Contains(c.Name)))
+                {
+                    var newColumnState = new ColumnState2(columnState.Name, isVisible: false, columnState.Width,
+                                columnState.SortPriority, columnState.DescendingSort, columnState.GroupingPriority);
```

Comment:
nit: strange indentaiton.  either '4' columns, or aligned with the rest of the args.